### DISCUSSION TITLE
Multithreaded server

### DISCUSSION THREAD

**Discussion by lukehutch** [serverpod/serverpod#654] [Q&A] [answered]
I want to start a number of isolates on the server to keep as many cores as possible busy serving requests. As far as I can tell, Serverpod uses just one main isolate for serving requests, but maybe I'm missing something.

* How can I spin up multiple Serverpod server instances in different isolates, for the same server code? Is there a way to spin up isolates automatically as needed? (I'm not talking about spinning up AWS instances on demand, I want to utilize all cores on a single machine.)
* Assuming these server instances all depend on a shared in-memory data structure, how can I pass updated copies of that data structure to all the running isolates, so that the data structure can be updated as needed by a writer isolate, without having to worry about readers reading inconsistent state?

**Comment by vlidholt** [comment, marked as answer]
Hi! Serverpod hasn't got support for multiple isolates yet. This is something we may want to do in the future. One solution you can currently do is to run multiple instances of Serverpod on a single machine on different ports, then pipe the ports through a load balancer.

Also, Dart naturally uses some multi-threading for handling e.g. for handling i/o. But I haven't done any tests on how large percentage of CPU is distributed that way.

To answer your second question, your different instances can communicate through Redis. There is an API for easily passing messages through pub/sub. You can find more about that here:
https://docs.serverpod.dev/concepts/streams

**Reply by lukehutch** [reply]
Is it possible to have one isolate handle many requests concurrently (i.e. asynchronously)? Or are requests handled sequentially, with a new request handled only when the previous request has returned a result? It would be great to at least be able to open say 8 different database connections, and be asynchronously serving requests, distributing all the requests across the database threads.

**Reply by vlidholt** [reply]
Yes, multiple requests are already handled concurrently in Serverpod. The database uses a pool of connections to talk with the database, so that too is working out of the box. :)

**Reply by lukehutch** [reply]
Awesome, nice work :) Can you please take a guess at the number of users that can be concurrently handled on a single Serverpod instance, with moderate database usage per user (queries with small result sets, hit every few seconds by each user)? Just looking for a rough ballpark, as there are mtiple factors that could obviously affect the result by an order of magnitude or two.

Or alternatively, let's say I want to support 100k concurrent active users, how many Serverpod instances would I likely need to spin up to prevent too much request latency? (Again, just a rough ballpark)

**Reply by vlidholt** [reply]
As you say, it very much depends on your application, plus which instances you are running on. But for simple tasks you should be able to handle thousands of requests per second on a single machine.
